### ISSUE TITLE
When my upstream has SSL certificate, Kong don't have to carry the client certificate access

### ISSUE THREAD

**Issue by yczhaoDevelop** [Mashape/kong#2633]
### Summary
My upstream has SSL certificate，but kong does not carry the certificate access.
I accessed through SOAP-UI  tool and  take a client certificate.The SOAP - UI return 401 code.
If I direct accessed through SOAP-UI  and  take a client certificate  without Kong.The SOAP - UI return 200 code.
I saw this solution client_ssl configuration, but does not meet our requirements.Because we have more than one upstream.We need to take different client certificate to different upstream
### Steps To Reproduce
the api info:
```
{
    "http_if_terminated": true,
    "id": "9fa11201-0c18-4f30-9a4f-bdc9ec63df1f",
    "upstream_read_timeout": 60000,
    "preserve_host": false,
    "created_at": 1497839855000,
    "upstream_connect_timeout": 60000,
    "upstream_url": "https://192.168.2.200/",
    "strip_uri": true,
    "https_only": false,
    "name": "nest-api",
    "retries": 5,
    "upstream_send_timeout": 60000,
    "hosts": [
        "nest.nestvision.com"
    ]
} 
```
the certificate info:
```
{
     "cert": "-----BEGIN CERTIFICATE-----",
     "id": "ce5e75aa-4c69-44e9-ab39-ffc7b8d72946",
     "snis": [
      "nest.nestvision.com"
      ],
      "key": "-----BEGIN RSA PRIVATE KEY-----",
      "created_at": 1497840564000
}
```
I accessed through SOAP-UI  tool and  take a client certificate
request:https://nest.nestvision.com:8443/ws/Patient
response:
```
HTTP/1.1 401 Unauthorized
Date: Mon, 19 Jun 2017 03:48:49 GMT
Content-Type: text/html; charset=UTF-8
Content-Length: 194
Connection: keep-alive
Server: nginx/1.8.0
X-Kong-Upstream-Latency: 27
X-Kong-Proxy-Latency: 0
Via: kong/0.10.3

<html>
<head><title>401 Authorization Required</title></head>
<body bgcolor="white">
<center><h1>401 Authorization Required</h1></center>
<hr><center>nginx/1.8.0</center>
</body>
</html>
```
### Additional Details & Logs

- Kong version (`0.10.3`)
- Kong accsess logs
```
 Kong configuration (register"POST /ws/Patient HTTP/1.1" 401 194 "-" "Apache-HttpClient/4.1.1 (java 1.5)"ed APIs/Plugins & configuration file)
```

**Comment by p0pr0ck5** [comment]
@yczhaoDevelop are you talking about using client SSL certificates? This functionality was added in Kong 0.10.3: https://github.com/Mashape/kong/blob/master/CHANGELOG.md#added

If not, can you help clarify exactly what you're looking for?

**Comment by yczhaoDevelop** [comment]
I mean the upstream is a bidirectional authentication upstream.The service side need to check the client's certificate.Now, the use of kong, but kong 0.10.3 currently does not support bidirectional authentication.Kong use openresty .Openresty currently does not support.
Now, I to develop openresty extension
lua-nginx-module/src/ngx_http_lua_ssl_certby.c 
```
int
ngx_http_lua_ffi_set_client_certificate(ngx_http_request_t *r,void *x509, char **err)
{
        ngx_ssl_conn_t    *ssl_conn;
        SSL_CTX  *ssl_ctx;
        X509_STORE *x509_store;


    if (r->connection == NULL || r->connection->ssl == NULL) {
        *err = "bad request";
        return NGX_ERROR;
    }

    ssl_conn = r->connection->ssl->connection;
    if (ssl_conn == NULL) {
        *err = "bad ssl conn";
        return NGX_ERROR;
    }
    ssl_ctx = SSL_get_SSL_CTX(ssl_conn);
    x509_store = X509_STORE_new();
    SSL_CTX_set_cert_store(ssl_ctx, x509_store);
    X509_STORE_add_cert(x509_store, x509);

    *err = NULL;
    return NGX_OK;
}


int
ngx_http_lua_ffi_set_client_verify(ngx_http_request_t *r,char **err)
{
        ngx_ssl_conn_t    *ssl_conn;


    if (r->connection == NULL || r->connection->ssl == NULL) {
        *err = "bad request";
        goto failed;
    }

    ssl_conn = r->connection->ssl->connection;

    if (ssl_conn == NULL) {
        *err = "bad ssl conn";
        goto failed;
    }
    SSL_set_verify(ssl_conn, SSL_VERIFY_PEER, NULL);

    return NGX_OK;
failed:
   ERR_clear_error();

   return NGX_ERROR;
}


void *
ngx_http_lua_ffi_parse_pem_client_certificate(const u_char *pem, size_t pem_len,
    char **err)
{
    BIO             *bio;
    X509            *x509;

    bio = BIO_new_mem_buf((char *) pem, (int) pem_len);
    if (bio == NULL) {
        *err = "BIO_new_mem_buf() failed";
        ERR_clear_error();
        return NULL;
    }

    x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
    if (x509 == NULL) {
        *err = "PEM_read_bio_X509_CRL() failed";
        BIO_free(bio);
        ERR_clear_error();
        return NULL;
    }
    return x509;

}

```
lua-resty-core/lib/ngx/ssl.lua

```
function _M.parse_pem_client_certificate(pem)
    local ca = C.ngx_http_lua_ffi_parse_pem_client_certificate(pem, #pem, errmsg)
    return ca
end



function _M.set_client_certificate(data)
    local r = getfenv(0).__ngx_req
    if not r then
        return error("no request found")
    end
    local rc = C.ngx_http_lua_ffi_set_client_certificate(r, data, errmsg)
    if rc == FFI_OK then
         return true
    end
    return nil
end



function _M.set_verify_client()
    local r = getfenv(0).__ngx_req
    if not r then
        return error("no request found")
    end

    local rc = C.ngx_http_lua_ffi_set_client_verify(r, errmsg)
    if rc == FFI_OK then
       return true
    end
    return nil, ffi_str(errmsg[0])
end

```
The above code is very simple.Use the above code in ssl_certificate_by_lua_block.The test is no problem.
I am a Java developer, not very familiar to C and LUA.I don't know the code above will there be other problems.If you know, can you tell me?
Now, I to development of kong, use the above code

**Comment by p0pr0ck5** [comment]
@yczhaoDevelop I suggest you propose a PR to the openresty project if you would like this to be included :) As noted, this is a duplicate of #2408, so closing this. Feel free to add your thoughts to that issue.
